File: src/org/cerion/symcalc/expression/function/core/Set.java
package org.cerion.symcalc.expression.function.core;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.VarExpr;

public class Set extends FunctionExpr {

    public Set(Expr... e) {
        super(FunctionType.SET, e);
    }

    @Override
    protected Expr evaluate() {
        VarExpr v = get(0).toVar();
        Expr e = get(1);

        getEnv().setVar(v.value(), e);

        return e;
    }
}

File: src/main/kotlin/function/core/Set.kt
package org.cerion.symcalc.function.core

import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.VarExpr
import org.cerion.symcalc.function.FunctionExpr

class Set(vararg e: Expr) : FunctionExpr(*e) {

    override fun evaluate(): Expr {
        val v = get(0) as VarExpr
        val e = get(1)

        env.setVar(v.value, e)

        return e
    }

    override fun validate() {
        validateParameterCount(2)
        validateParameterType(0, ExprType.VARIABLE)
    }
}
